### PR TITLE
Issue 116. Remove closed connections from pool.

### PR DESCRIPTION
**Pull request by runtel**
Hi!
Pease check commit at https://github.com/aio-libs/aiopg/issues/116

Thanks

### CODE REVIEW

File: aiopg/pool.py [hunk 1]
```diff
@@ -176,6 +176,17 @@ def _acquire(self):
 
     @asyncio.coroutine
     def _fill_free_pool(self, override_min):
+        # iterate over free connections and remove timeouted ones
+        n, free = 0, len(self._free)
+        print(11111)
```
Comment (popravich) [inline aiopg/pool.py]: remove `print`

File: tests/test_pool.py [hunk 2]
```diff
@@ -527,3 +527,34 @@ def sleep(conn):
...
+@pytest.mark.run_loop
+def test_drop_connection_if_timedout(create_pool, loop):
+
+    @asyncio.coroutine
+    def _set_global_conn_timeout(pool, t):
+        # Set timeout for client connection
+        # https://www.postgresql.org/docs/current/static/runtime-config-client.html
+        conn = yield from pool.acquire()
+        cur = yield from conn.cursor()
+        yield from cur.execute('set statement_timeout to %s;' % t)
```
Comment (jettify) [inline tests/test_pool.py]: I am not sure that this is correct, `statement_timeout` is timeout for statement execution time, not the timeout for idle connection.

**Comment by runtel** [comment]
I have been created new test. 

```python
@pytest.mark.run_loop
def test_drop_connection_if_timedout(create_pool, loop):

    @asyncio.coroutine
    def _kill_connectios(pool):
        # Drop all connections on server
        conn = yield from pool.acquire()
        cur = yield from conn.cursor()
        try:
            yield from cur.execute("""WITH inactive_connections_list AS (
            SELECT pid FROM  pg_stat_activity)
            SELECT pg_terminate_backend(pid) FROM inactive_connections_list""")
        except DatabaseError:
            pass
        pool.release(conn)
        conn.close()

    pool = yield from create_pool(minsize=3, maxsize=5)
    yield from _kill_connectios(pool)
    yield from asyncio.sleep(1, loop=loop)
    conn = yield from pool.acquire()
    cur = yield from conn.cursor()
    yield from cur.execute('SELECT 1;')
    pool.release(conn)
    conn.close()
    pool.close()
    yield from pool.wait_closed()
```

Server dropped all connections and than this pool renewed connections.

**Comment by jettify** [comment]
@runtel test looks good! Just couple simple improvements are required. Could you update your PR with new test?
1) in `_kill_connections` you do not have to create pool, just create connection directly with `make_connection` fixture
2) why you did you wrap `cur.execute` in try/except?

**Comment by runtel** [comment]
@jettify 
Update for unitest. Is correct?
```python

@pytest.mark.run_loop
def test_drop_connection_if_timedout(make_connection, pg_params,
                                     create_pool, loop):

    @asyncio.coroutine
    def _kill_connectios():
        # Drop all connections on server
        conn = yield from make_connection()
        cur = yield from conn.cursor()
        try:
            yield from cur.execute("""WITH inactive_connections_list AS (
            SELECT pid FROM  pg_stat_activity)
            SELECT pg_terminate_backend(pid) FROM inactive_connections_list""")
        except DatabaseError:
            # Pass psycopg2.DatabaseError: server closed the connection unexpectedly
            pass
        conn.close()

    pool = yield from create_pool()
    yield from _kill_connectios()
    yield from asyncio.sleep(1, loop=loop)
    conn = yield from pool.acquire()
    cur = yield from conn.cursor()
    yield from cur.execute('SELECT 1;')
    pool.release(conn)
    conn.close()
    pool.close()
    yield from pool.wait_closed()
```

**Comment by jettify** [comment]
Test case looks good. One question, have you tried to to run it without try /except? I saw somewhere that current connection can not be closed... but probably I missed something.

**Comment by runtel** [comment]
@jettify I can not remove try/except. 

```python

make_connection = <function make_connection.<locals>.go at 0x7f5298a57950>, pg_params = {'database': 'postgres', 'host': '127.0.0.1', 'password': 'mysecretpassword', 'port': 44761, ...}, create_pool = <function create_pool.<locals>.go at 0x7f5298a57048>
loop = <_UnixSelectorEventLoop running=False closed=False debug=False>

    @pytest.mark.run_loop
    def test_drop_connection_if_timedout(make_connection, pg_params,
                                         create_pool, loop):

        @asyncio.coroutine
        def _kill_connectios():
            # Drop all connections on server
            conn = yield from make_connection()
            cur = yield from conn.cursor()
            # try:
            yield from cur.execute("""WITH inactive_connections_list AS (
                SELECT pid FROM  pg_stat_activity)
                SELECT pg_terminate_backend(pid) FROM inactive_connections_list""")
            # except DatabaseError:
            #     pass
            conn.close()

        pool = yield from create_pool()
>       yield from _kill_connectios()

tests/test_pool.py:551:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_pool.py:545: in _kill_connectios
    SELECT pg_terminate_backend(pid) FROM inactive_connections_list""")
aiopg/cursor.py:113: in execute
    yield from self._conn._poll(waiter, timeout)
aiopg/connection.py:193: in _poll
    yield from asyncio.wait_for(self._waiter, timeout, loop=self._loop)
/usr/local/lib/python3.5/asyncio/tasks.py:386: in wait_for
    return fut.result()
/usr/local/lib/python3.5/asyncio/futures.py:274: in result
    raise self._exception
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

weak_self = <weakref at 0x7f5298a443b8; to 'Connection' at 0x7f5298f877b8>

    @staticmethod
    def _ready(weak_self):
        self = weak_self()
        if self is None:
            return

        waiter = self._waiter

        try:
>           state = self._conn.poll()
E           psycopg2.DatabaseError: server closed the connection unexpectedly
E           	This probably means the server terminated abnormally
E           	before or while processing the request.

```

**Comment by runtel** [comment]
I update PR?

**Comment by jettify** [comment]
Please do this :)

**Comment by jettify** [comment]
Thanks!

**Comment by runtel** [comment]
you're welcome)

**Pull request merged**